aten/src/ATen/native/mkldnn/xpu/detail/Utils.h

```diff
@@ -0,0 +1,376 @@
+  TORCH_CHECK(
+      3 == ndim || 4 == ndim || 5 == ndim,
+      "ndim must be 3, 4 or 5 when get cl tag");
+  if (3 == ndim) {
+    return at::MemoryFormat::Contiguous;
+  } else if (5 == ndim) {
+    return at::MemoryFormat::ChannelsLast3d;
+  } else {
+    return at::MemoryFormat::ChannelsLast;
+  }
+}
+
+static inline bool binary_valid(
+    const at::Tensor& self,
+    const at::Tensor& other,
+    bool is_fusion = false) {
+  // FIXME: update onednn
+  if (self.sizes() != other.sizes() &&
+      !is_broadcast_from_other_to_self(self, other))
+    return false;
+
+  /* If the following conditions are satisfied, then oneDNN path will be
+     selected:
+     * 1. self and other should be xpu tensor and be defined.
+     * 2. self or other should not be scalar (wrapped tensor).
+     * 3. dim of self and other should be equal and must be larger than 0 and
+     smaller than 7.
+     * 4. the datatype should be supported by oneDNN primitive.
+     * 5. self and other should be in the same datatype.
+     * 6. self and other should be contiguous or channel-last contiguous.*/
+
+
+  // 1. self and other should be xpu tensor and be defined.
+  if ((!self.defined()) || (!other.defined()) || (!self.is_xpu()) ||
+      (!other.is_xpu()))
+    return false;
+
+  // 2. self or other should not be scalar (wrapped tensor).
+  if (is_wrapped_number(self) || is_wrapped_number(other))
+    return false;
+
+  // 3. dim of self and other should be equal and must be larger than 0 and
+  // smaller than 7.
+  if ((self.dim() <= 0) || (other.dim() <= 0) || (self.dim() != other.dim()) ||
+      (self.dim() > 6) || (other.dim() > 6))
+    return false;
+
+  // 4. the datatype should be supported by oneDNN primitive.
+  switch (self.scalar_type()) {
+    case at::ScalarType::Char:
+      break;
+    case at::ScalarType::Byte:
+      break;
+    case at::ScalarType::Half:
+      break;
+    case at::ScalarType::Float:
+      break;
+    case at::ScalarType::BFloat16:
+      break;
+    default:
+      return false;
+  };
+
+  // 5. datatype check
+  if (is_fusion) {
+    // for fusion case, the fusion can be performed on scalar_type or Float
+    // datatype.
+    if (self.scalar_type() != other.scalar_type() &&
+        other.scalar_type() != at::ScalarType::Float) {
+      return false;
+    }
+  } else {
+    if (self.scalar_type() != other.scalar_type()) {
+      // for non-fusion case: self and other should be in the same datatype.
+      return false;
+    }
+  }
+
+  // 6. self and other should be contiguous or channel-last contiguous.
+  const auto ndim = self.ndimension();
+  auto cl_tag = at::MemoryFormat::ChannelsLast;
+  if (3 == ndim || 4 == ndim || 5 == ndim) {
+    cl_tag = get_cl_tag_by_ndim(ndim);
+  }
+  if ((self.is_contiguous() && other.is_contiguous()) ||
+      (self.is_contiguous(cl_tag) && other.is_contiguous(cl_tag)))
+    return true;
+  return false;
+}
+
+enum MEMORY_LAYOUT_FOR_CONV {
```

Comment:
Is it necessary to define in integration code?

Reply:
removed